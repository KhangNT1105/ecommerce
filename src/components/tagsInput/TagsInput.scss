@import "styles/variables";

.input-tag-container {
  .input-tag {
    background: $white;
    border: 1px solid $border-input;
    border-radius: 2px;
    display: flex;
    flex-wrap: wrap;
    padding: 5px 5px 0;

    .input-tag-list {
      display: inline-flex;
      flex-wrap: wrap;
      margin: 0;
      padding: 0;
      width: 100%;

      li {
        align-items: center;
        background: $grey-1;
        border-radius: 2px;
        color: $black;
        display: flex;
        font-weight: 400;
        margin-bottom: 5px;
        margin-right: 5px;
        padding: 5px 10px;
      }

      .tag-button {
        align-items: center;
        border: none;
        border-radius: 50%;
        background: $grey-1;
        font-size: 18px;
        height: 16px;
        justify-content: center;
        line-height: 0;
        margin-left: 4px;
        padding: 0;
        width: 16px;
        font-weight: bold;
        overflow: hidden;
      }

      .input-box {
        background: none;
        flex-grow: 1;
        padding: 0;

        input {
          border: none;
          width: 100%;
        }

        input:invalid {
          box-shadow: none;
        }
      }
    }
  }

  .dropdown {
    width: 100%;
    top: 100%;
    left: 0;
    background: $white;

    .suggestions {
      list-style: none;
      padding-left: 0;
      box-shadow: 0 1px 6px $border-filter, 0 1px 4px $border-filter;

      li {
        border-bottom: 1px solid $grey-1;
        padding: 0.6em;
        cursor: pointer;
        user-select: none;
      }

      li:hover {
        background-color: $grey-1;
      }

      .focused {
        transition: backgroundcolor linear 0.2s;
        background-color: $grey-1;
      }
    }
  }
}
