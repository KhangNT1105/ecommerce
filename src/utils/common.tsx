import React from 'react';
import { Denomination } from 'Props.d';
import { assign, clone, map, reduce, isObjectLike } from 'lodash';
import { validateRequired, validateEmail } from 'utils/validation';
import { ACTIVATE, ENTITY_STATUS_FROM_SERVER, DEACTIVATE, DELETE, APPROVE, REJECT } from '../constants';
import { Genre, SubGenre } from 'stores/ConfigurationStore/configurationType.d';

export const checkRenderDoubleValue = (value1?: string | number, value2?: string | number, value3?: any | '') => {
  let minus = '';
  if (value1 && value2) {
    minus = ' - ';
  }
  return (
    <>
      {value1} {value3 && <b>{value3}</b>}
      {minus}
      {value2}
    </>
  );
};

export type SelectOptionProps = {
  id: string;
  value: string;
  label: string;
};

type DataProps = {
  [key: string]: any;
};

export type SelectOptionKeyMappingProps = {
  key: string;
  value: string;
  label: string;
};

export const renderSelectOptions = (data: DataProps[], keyMapping?: SelectOptionKeyMappingProps) => {
  return (
    <>
      <option value="">Select</option>
      {data.map((item) => {
        if (keyMapping && Object.entries(keyMapping).length > 0) {
          return (
            <option key={item[keyMapping.key]} value={item[keyMapping.value]}>
              {item[keyMapping.label]}
            </option>
          );
        }
        return (
          <option key={item.id} value={item.value}>
            {item.label}
          </option>
        );
      })}
    </>
  );
};

export const checkRenderArray = (items: string[]) => {
  return <span>{items.join(', ')}</span>;
};

export const getListPrices = (denominations: Denomination[]) => {
  const listPrices = denominations.map((item) => item.price);
  return listPrices.sort((a, b) => a - b);
};
export const getPrices = (productDetail: any) => {
  if (productDetail.valueType === 'RANGE') {
    return checkRenderDoubleValue(productDetail.minPrice, productDetail.maxPrice);
  }
  if (productDetail.valueType === 'SLAB') {
    const values = getListPrices(productDetail.denominations);
    return values.join(', ');
  }
};

export interface IPagingModelOption {
  pageIndex: number;
  pageSize: number;
}

export interface IPagingModelResult<T> {
  items: T[];
  hasPrevPage: boolean;
  hasNextPage: boolean;
  prevPage: number | null;
  nextPage: number | null;
  size: number;
  page: number;
  totalItems: number;
  totalPages: number;
}

export function getPaginationResponse<T>(
  pagingModel: IPagingModelOption,
  result: T[],
  totalItems: number
): IPagingModelResult<T> {
  const totalPages = Math.floor((totalItems + pagingModel.pageSize - 1) / pagingModel.pageSize);
  const pagingModelResult: IPagingModelResult<T> = {
    items: result,
    totalItems,
    totalPages,
    hasPrevPage: pagingModel.pageIndex > 0,
    hasNextPage: pagingModel.pageIndex < totalPages - 1,
    prevPage: pagingModel.pageIndex > 0 ? pagingModel.pageIndex - 1 : null,
    nextPage: pagingModel.pageIndex < totalPages - 1 ? pagingModel.pageIndex + 1 : null,
    size: pagingModel.pageSize,
    page: pagingModel.pageIndex
  };
  return pagingModelResult;
}

export function replaceSpecialCharacters(str: string, character: string) {
  return str.replace(/(?:\.(?![^.]+$)|[^\w.])+/g, character);
}

export const classConcat = (classes: string[]): string => {
  if (classes.length) {
    return classes.filter((className) => !!className).join(' ');
  }
  return '';
};

export const recursiveGetFieldName = (obj: object, entityField: string, fieldNameCombine?: string): any => {
  let listFieldConfig: any;
  listFieldConfig = reduce(
    obj,
    (result, value, key) => {
      let cloneResult = clone(result);
      let localFieldName: any = clone(fieldNameCombine);
      if (key === entityField) {
        map(value, (item: any, k) => {
          const getFieldName = `${localFieldName}.${key}.${k}`;

          // check whether any validation
          if (item.validationRules && item.validationRules.length > 0) {
            item.validationRules.map((itemValidate: any, index: number) => {
              if (itemValidate.rule === 'isRequired') {
                item.validationRules[index].validator = validateRequired;
              }
              if (itemValidate.rule === 'isEmail') {
                item.validationRules[index].validator = validateEmail;
              }

              return itemValidate;
            });
          }

          // insert key, value into list field
          return (cloneResult = {
            ...cloneResult,
            [getFieldName]: item
          });
        }); // reset field name

        localFieldName = '';
      } else if (isObjectLike(value)) {
        if (!localFieldName) {
          localFieldName = key;
        } else {
          localFieldName += `.${key}`;
        } // recall recursive fn if not be found the expected entityField

        return assign(cloneResult, recursiveGetFieldName(value, entityField, localFieldName));
      }

      listFieldConfig = {
        ...listFieldConfig,
        ...cloneResult
      };

      return listFieldConfig;
    },
    {}
  );

  return listFieldConfig;
};

export type ObjectFieldType = {
  [field: string]: string | object | string[] | object[] | any;
};

export const trimAllValueWithObjectField = (objField: ObjectFieldType) => {
  if ((!Array.isArray(objField) && typeof objField !== 'object') || objField === null) {
    return objField;
  }
  return Object.keys(objField).reduce(
    (acc: ObjectFieldType, key) => {
      acc[key] =
        typeof objField[key] === 'string'
          ? (objField[key] as string).trim()
          : trimAllValueWithObjectField(objField[key]);
      return acc;
    },
    Array.isArray(objField) ? [] : {}
  );
};

const listShowBtnActivate = [
  ENTITY_STATUS_FROM_SERVER.INACTIVATE,
  ENTITY_STATUS_FROM_SERVER.DRAFT,
  ENTITY_STATUS_FROM_SERVER.UPDATED,
  ENTITY_STATUS_FROM_SERVER.ACTIVATE
];

const listShowBtnSMActivate = [
  ENTITY_STATUS_FROM_SERVER.INACTIVATE,
  ENTITY_STATUS_FROM_SERVER.DRAFT,
  ENTITY_STATUS_FROM_SERVER.UPDATED
];

const listShowBtnSMDelete = [ENTITY_STATUS_FROM_SERVER.DRAFT, ENTITY_STATUS_FROM_SERVER.INACTIVATE];

const listShowBtnDeactivate = [ENTITY_STATUS_FROM_SERVER.ACTIVATE, ENTITY_STATUS_FROM_SERVER.UPDATED];

const listShowBtnDelete = [ENTITY_STATUS_FROM_SERVER.INACTIVATE, ENTITY_STATUS_FROM_SERVER.DRAFT];

export const checkEntityActionNonApproval = (action: string, entityStatus: string) => {
  switch (action) {
    case ACTIVATE:
      return listShowBtnActivate.includes(entityStatus);
    case DEACTIVATE:
      return listShowBtnDeactivate.includes(entityStatus);
    case DELETE:
      return listShowBtnDelete.includes(entityStatus);
    default:
      return true;
  }
};

export const checkSMEntityActionNonApproval = (action: string, entityStatus: string) => {
  switch (action) {
    case ACTIVATE:
      return listShowBtnSMActivate.includes(entityStatus);
    case DEACTIVATE:
      return listShowBtnDeactivate.includes(entityStatus);
    case DELETE:
      return listShowBtnSMDelete.includes(entityStatus);
    default:
      return true;
  }
};

export const getAllSubGenres = (genres: Genre[]) =>
  genres
    .map(({ children }) => children)
    .reduce((currentValue, subGenres) => {
      if (subGenres) {
        return [...currentValue, ...subGenres];
      }
      return currentValue;
    }, []);

export const getAllMicroGenres = (subGenres: SubGenre[]) =>
  subGenres
    .map(({ children }) => children)
    .reduce((currentValue, microGenre) => {
      if (microGenre) {
        return [...currentValue, ...microGenre];
      }
      return currentValue;
    }, []);

const listShowBtnActivateWithApproval = [...listShowBtnActivate, ENTITY_STATUS_FROM_SERVER.REJECT];
const listShowBtnDeactivateWithApproval = [...listShowBtnDeactivate];
const listShowBtnDeleteWithApproval = [...listShowBtnDelete];
const listShowBtnApproveWithApproval = [ENTITY_STATUS_FROM_SERVER.APPROVAL_PENDING];
const listShowBtnRejectWithApproval = [ENTITY_STATUS_FROM_SERVER.APPROVAL_PENDING];

export const checkEntityActionApproval = (action: string, entityStatus: string) => {
  switch (action) {
    case ACTIVATE:
      return listShowBtnActivateWithApproval.includes(entityStatus);
    case DEACTIVATE:
      return listShowBtnDeactivateWithApproval.includes(entityStatus);
    case DELETE:
      return listShowBtnDeleteWithApproval.includes(entityStatus);
    case APPROVE:
      return listShowBtnApproveWithApproval.includes(entityStatus);
    case REJECT:
      return listShowBtnRejectWithApproval.includes(entityStatus);
    default:
      return true;
  }
};

export function createMarkup() {
  return {
    __html: `<div class="zalo-chat-widget" data-oaid="927003371386666481" data-welcome-message="Rất vui khi được hỗ trợ bạn!" data-autopopup="0" data-width="450" data-height="450"></div>
    `
  };
}
