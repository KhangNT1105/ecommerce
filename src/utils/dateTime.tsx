import {
  format,
  parse,
  getTime,
  subDays,
  subMonths,
  startOfWeek,
  endOfWeek,
  startOfMonth,
  endOfMonth,
  startOfToday,
  endOfToday,
  startOfDay,
  endOfDay
} from 'date-fns';

import { DATE_FORMAT, DATE_TIME_FULL_FORMAT } from 'constants/index';

export type DateRangeType = { from: string | null; to: string | null };

export const formatDate = (timestamp: string) => {
  return format(parse(timestamp, 'T', new Date()), DATE_FORMAT);
};

export type DateRangeProps = {
  dateFrom: string | null;
  dateTo: string | null;
};

/**
 * get timestamp from a string
 * @param: value string
 */
export const getTimestamp = (value: string): number => {
  return getTime(parse(value, DATE_FORMAT, new Date()));
};

export const getUTCDateFn = (date: Date) => {
  return new Date(
    date.getUTCFullYear(),
    date.getUTCMonth(),
    date.getUTCDate(),
    date.getUTCHours(),
    date.getUTCMinutes(),
    date.getUTCSeconds()
  );
};

/**
 * Description: combine day from DatePicker and local time then convert to UTC Date
 * @param date: from date picker
 * @return: UTC Date
 */
export const convertDatePickerToUTC = (date: Date) => {
  const today = new Date();
  const combineDateTime = new Date(
    date.getFullYear(),
    date.getMonth(),
    date.getDate(),
    today.getHours(),
    today.getMinutes(),
    today.getSeconds()
  );

  return getUTCDateFn(combineDateTime);
};

/**
 * Description: convert HH:MM:SS string to seconds
 * @param dateString
 * @param second
 */
export const convertTimeStringToMillisecond = (dateString: String) => {
  return (dateString as any).split(':').reduce((acc: any, time: any) => 60 * acc + +time) * 1000;
};

export const convertDateFromToUTC = (date: Date) => {
  const combineDateTime = new Date(date.getFullYear(), date.getMonth(), date.getDate());

  return getUTCDateFn(startOfDay(combineDateTime));
};

export const convertDateToToUTC = (date: Date) => {
  const combineDateTime = new Date(date.getFullYear(), date.getMonth(), date.getDate());

  return getUTCDateFn(endOfDay(combineDateTime));
};

export const getToday = (): DateRangeProps => {
  const dateFrom = getUTCDateFn(startOfToday());
  const dateTo = getUTCDateFn(endOfToday());

  return {
    dateFrom: format(dateFrom, DATE_TIME_FULL_FORMAT),
    dateTo: format(dateTo, DATE_TIME_FULL_FORMAT)
  };
};

export const getLast7days = (): DateRangeProps => {
  const dateTo = format(getUTCDateFn(endOfToday()), DATE_TIME_FULL_FORMAT);
  const dateFrom = format(getUTCDateFn(subDays(startOfToday(), 7)), DATE_TIME_FULL_FORMAT);

  return { dateFrom, dateTo };
};

export const getLast30days = (): DateRangeProps => {
  const dateTo = format(getUTCDateFn(endOfToday()), DATE_TIME_FULL_FORMAT);
  const dateFrom = format(getUTCDateFn(subDays(startOfToday(), 30)), DATE_TIME_FULL_FORMAT);

  return { dateFrom, dateTo };
};

export const getLastWeek = (): DateRangeProps => {
  const dateFrom = format(getUTCDateFn(startOfWeek(subDays(startOfToday(), 7))), DATE_TIME_FULL_FORMAT);
  const dateTo = format(getUTCDateFn(endOfWeek(subDays(endOfToday(), 7))), DATE_TIME_FULL_FORMAT);

  return { dateFrom, dateTo };
};

export const getLastMonth = (): DateRangeProps => {
  const dateFrom = format(getUTCDateFn(startOfMonth(subMonths(startOfToday(), 1))), DATE_TIME_FULL_FORMAT);
  const dateTo = format(getUTCDateFn(endOfMonth(subMonths(endOfToday(), 1))), DATE_TIME_FULL_FORMAT);

  return { dateFrom, dateTo };
};

export const convertMillisecondToHHMMSS = (milliseconds: number) => {
  if (!milliseconds) return '';
  const date = new Date(milliseconds);
  const hh = date.getUTCHours();
  const mm = date.getUTCMinutes();
  const ss = date.getSeconds();
  const hours = hh < 10 ? `0${hh}` : `${hh}`;
  const minutes = mm < 10 ? `0${mm}` : `${mm}`;
  const seconds = ss < 10 ? `0${ss}` : `${ss}`;

  return `${hours}:${minutes}:${seconds}`;
};
