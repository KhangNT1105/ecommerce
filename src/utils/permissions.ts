import { matchPath } from 'react-router';
import { FieldType, PermissionFieldsType, FormPermissionsType } from '../components/roleFormField/RoleFormFields.d';
import { BUTTONS_VISIBILITY, PERMISSION_PREFIX, PAGE_ACCESSIBILITY } from '../constants/enum';
/**
 * Description: check user permissions have spec permissions
 * @param permissions: array of user permissions
 * @param modes: array of PERMISSION_PREFIX
 * Results: boolean
 */
export const hasModes = (permissions: string[], modes: string[]) => {
  const foundMode = permissions.filter((item) => {
    const includes = modes.filter((mode) => {
      return item === mode;
    });
    return includes.length > 0;
  });
  return foundMode.length === modes.length;
};

/**
 * Description: check user permissions only have 1 permission
 * @param permissions: array of user permissions
 * @param mode: PERMISSION_PREFIX
 * Results: boolean
 */
export const hasOnlyMode = (permissions: string[], mode: string) => {
  const viewMode = permissions.filter((item) => {
    return item === mode;
  });
  return permissions.length === 1 && viewMode.length === 1;
};

/**
 * Description: generate new structure data:
 * Params:
 *    @param formPermissions: array of permissions and element name (id)
 *    @param userPermissions: array of permissions of logged in users
 *
 * Results:
 * {
 *  btnSave: 'show',
 *  btnCancel: 'NA',
 *  ...
 * }
 */
export const reGeneratePermissionFields = (formPermissions: FormPermissionsType[], userPermissions: string[]) => {
  const permissionFields: PermissionFieldsType = formPermissions
    .filter(({ permissions }) => {
      return userPermissions.includes(permissions);
    })
    .map((formPermissions: FormPermissionsType) => {
      const { permissions, ...restPermissions } = formPermissions;
      return restPermissions;
    });

  const fieldsWithPermissions = permissionFields.reduce(
    (accumulator: FieldType, currentValue: FieldType, index: number) => {
      const fields = {
        ...accumulator
      };
      Object.keys(currentValue).map((operation: string) => {
        if (currentValue[operation] === BUTTONS_VISIBILITY.SHOW) {
          fields[operation] = currentValue[operation];
        }
        return true;
      });
      return {
        ...fields
      };
    },
    {}
  );
  return fieldsWithPermissions;
};
/**
 * Description: get accessibility of page with specification page
 * Params:
 *    @param formPermissions: array of permissions of page
 *    @param userPermissions: array of permissions of logged in users
 *    @param strPermission: permission
 *
 * Results:
 * pageAccessibility: boolean
 */
export const checkPageAccessibility = (
  formPermissions: FormPermissionsType[],
  userPermissions: string[],
  strPermission: string
) => {
  const pagePermissions = reGeneratePermissionFields(formPermissions, userPermissions);
  return pagePermissions[strPermission] === PAGE_ACCESSIBILITY.SHOW;
};

/**
 * Description: get accessibility of page with specification entity
 * Params:
 *    @param path: current path of site
 *    @param entityName: entity
 *    @param formPermissionsCreate: array of permissions of create page
 *    @param formPermissionsEdit: array of permissions of edit page
 *    @param formPermissionsList: array of permissions of list page
 *    @param pathCreate: route string of create page
 *    @param pathEdit: route string of edit page
 *    @param userPermissions: array of permissions of logged in users
 *
 * Results:
 * pageAccessibility: boolean
 */
export const getPageAccessibility = (
  path: string,
  entityName: string,
  formPermissionsCreate: FormPermissionsType[],
  formPermissionsEdit: FormPermissionsType[],
  formPermissionsList: FormPermissionsType[],
  pathCreate: string,
  pathEdit: string,
  userPermissions: string[]
) => {
  let pageAccessibility = true;
  if (matchPath(path, { path: pathCreate, exact: true })) {
    pageAccessibility = checkPageAccessibility(
      formPermissionsCreate,
      userPermissions,
      `${entityName}_${PERMISSION_PREFIX.CREATE.toLowerCase()}`
    );
  } else if (matchPath(path, { path: pathEdit, exact: true })) {
    pageAccessibility = checkPageAccessibility(
      formPermissionsEdit,
      userPermissions,
      `${entityName}_${PERMISSION_PREFIX.EDIT.toLowerCase()}`
    );
  } else {
    pageAccessibility = checkPageAccessibility(formPermissionsList, userPermissions, `${entityName}`);
  }
  return pageAccessibility;
};

/**
 * Description: check page is accessible by entity
 * Params:
 *    @param page: page name
 *    @param path: current path of site
 *    @param userPermissions: array of permissions of logged in users
 *
 * Results:
 * pageAccessibility: boolean
 */
export const checkPagePermissisons = (page: string, path: string, userPermissions: string[]) => {
  // let pagePermissions: { [permission: string]: string };
  const pageAccessibility = true;

  return pageAccessibility;
};

/**
 * Description: get form permissions of list page by entity
 * Params:
 *    @param entity: entity name
 *
 * Results:
 * formPermissions: FormPermissionsType[]
 */
export const getFormPermissionsPageList = (entity: string) => {};
