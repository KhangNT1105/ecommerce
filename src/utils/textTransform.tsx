export const upperCaseFirst = (str: String = '') => {
  return str.toLowerCase().replace(/^\w/, (chars) => chars.toUpperCase());
};
