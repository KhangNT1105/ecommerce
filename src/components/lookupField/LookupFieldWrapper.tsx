import React, { useEffect, useState } from 'react';
import axios from 'axios';
import { Spinner } from 'reactstrap';

import API from 'api';

import LookupFieldTags from 'components/lookupField/lookupFieldChild/LookupFieldTags';
import { Params } from 'components/tagsInput/TagsInput.d';
import {
  WrapperPropsOther,
  ILookupFieldProps,
  TagsInputPropsOther,
  TagElementOther
} from 'components/lookupField/LookupFieldType.d';

const LookupFieldWrapper: React.FC<TagsInputPropsOther & ILookupFieldProps & WrapperPropsOther> = ({
  errorText,
  endpoint,
  fieldName,
  readonly,
  shouldClearTags,
  single,
  selectedTags,
  sort,
  suggestions = [],
  touch,
  setSelectedTags,
  clearTags,
  mappingFn,
  shouldAddNew,
  renderElement,
  isShowSuggestion,
  shouldHaveStatus
}) => {
  const [state, actions] = useState({
    isSearching: false,
    params: shouldHaveStatus ? { status: ['activated', 'updated'] } : {}
  });

  const [data, setData] = useState(suggestions);
  const [currentParams, setCurrentParams] = useState<Params | undefined>({});
  const [loading, setLoading] = useState(false);
  const [isMapping, setIsMapping] = useState(true);

  useEffect(() => {
    // check if mappingFn exist
    if (isMapping && mappingFn) {
      setData(mappingFn(data));
      setIsMapping(false);
    }
  }, [mappingFn, setData, data, isMapping]);

  useEffect(() => {
    if (endpoint) {
      const cancelTokenSource = axios.CancelToken.source();
      // check endpoint existing, call api
      const fetchData = async function () {
        try {
          setLoading(true);
          const response = await API({
            url: endpoint,
            params: currentParams,
            method: 'GET',
            headers: {},
            data: {},
            cancelTokenSource
          });

          let respContent = response.content;
          // check if mappingFn exist
          if (mappingFn) {
            respContent = mappingFn(response);
          }
          return setData(respContent);
        } catch (error) {
          throw error;
        } finally {
          setLoading(false);
        }
      };

      if (currentParams!.text) {
        fetchData();
      }

      return () => {
        cancelTokenSource && cancelTokenSource.cancel();
      };
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [
    endpoint,
    currentParams
    //  mappingFn // this makes the endpoint re-invokes when other fields input changes
  ]);

  useEffect(() => {
    if (state.isSearching) {
      setCurrentParams(state.params);

      if (!endpoint) {
        let dataFiltered: TagElementOther[] = [];
        if (currentParams!.text) {
          const regex = new RegExp(`${currentParams!.text}`, 'gi');
          dataFiltered = suggestions.filter(
            (suggestion) => !!suggestion[fieldName] && suggestion[fieldName].search(regex) > -1
          );
        }
        setData(dataFiltered);
      }
    }
  }, [setCurrentParams, state, currentParams, endpoint, suggestions, selectedTags, fieldName]);

  useEffect(() => {
    if (suggestions && suggestions.length) {
      setData([...suggestions]);
    }
  }, [suggestions]);

  const handleInputChange = (query: string) => {
    // check input text length must be greater than 1 to call api
    if (query.length > 1) {
      const newParam = {
        ...state.params,
        text: query.trim(),
        sort
      };
      actions({ isSearching: true, params: newParam });
    }
  };

  return (
    <LookupFieldTags
      errorText={errorText}
      fieldName={fieldName}
      isLoading={loading}
      readonly={readonly}
      selectedTags={selectedTags}
      single={single}
      suggestions={data}
      shouldClearTags={shouldClearTags}
      touch={touch}
      customLoader={<Spinner size="sm" color="secondary" />}
      clearTags={clearTags}
      setSelectedTags={setSelectedTags}
      handleInputChange={handleInputChange}
      shouldAddNew={shouldAddNew}
      renderElement={renderElement}
      isShowSuggestion={isShowSuggestion}
    />
  );
};

export default LookupFieldWrapper;
