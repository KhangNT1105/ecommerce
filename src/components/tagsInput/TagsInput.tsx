import React, { useState, useEffect, useRef } from 'react';
import { Col, FormFeedback } from 'reactstrap';
import { useTranslation } from 'react-i18next';

import Tag from './tag/Tag';
import InputTag from './inputTag/InputTag';
import Suggestions from './suggestions/Suggestions';

import './TagsInput.scss';
import { TagsInputProps, TagElement } from './TagsInput.d';
import { KEYS } from '../../constants/keyCode.type';
import { validateEmail } from 'utils/validation';

const TagsInput: React.FC<TagsInputProps> = ({
  readonly,
  suggestions,
  isLoading = false,
  customLoader,
  handleInputChange,
  getSelectedTags,
  shouldClearTags,
  clearTags
}) => {
  const [tagsState, setTagsState] = useState<TagElement[]>([]);
  const [focusedSuggestion, setFocusedSuggestion] = useState(0);
  const [filteredSuggestion, setFilteredSuggestion] = useState([]);
  const [isFocus, setFocus] = useState(false);
  const [inputValue, setInputValue] = useState('');
  const inputRef = useRef<HTMLInputElement>(null);
  const [errorMessage, setErrorMessage] = useState('');

  const { t } = useTranslation();

  useEffect(() => {
    if (shouldClearTags) {
      setTagsState([]);
    }
  }, [shouldClearTags]);

  useEffect(() => {
    const filterSuggestionFunction = () => {
      return suggestions
        ? suggestions.filter((suggestion: TagElement | any) => !tagsState.find((tag) => suggestion.id === tag.id))
        : suggestions;
    };

    setFilteredSuggestion(filterSuggestionFunction());
  }, [suggestions, tagsState]);

  const removeTag = (index: number) => {
    const newTags: TagElement[] = [...tagsState];
    newTags.splice(index, 1);
    setTagsState(newTags);
    getSelectedTags && getSelectedTags(newTags);
    inputRef.current?.focus();
  };

  const handleKeyDown = (e: React.KeyboardEvent<HTMLInputElement>) => {
    const val = e.currentTarget.value;
    const key = e.keyCode;

    let newTags: TagElement[];
    setInputValue(val);
    setErrorMessage('');

    if (key === KEYS.ENTER && val) {
      if (!filteredSuggestion?.length) {
        if (validateEmail(val)) {
          const mappingTag = {
            id: val,
            email: val,
            addedDate: ''
          };
          if (checkDuplicateTag(mappingTag)) {
            setErrorMessage(t('EXISTING_VALUE_ERROR', { value: mappingTag.email }));
            return;
          }

          newTags = [...tagsState, mappingTag];
        } else {
          setErrorMessage(t('FORM_ERROR_REQUIRED_FIELD_EMAIL_INVALID'));
          return;
        }
      } else {
        const newFocus = focusedSuggestion < 0 ? 0 : focusedSuggestion;
        if (checkDuplicateTag(filteredSuggestion[newFocus])) return;
        newTags = [...tagsState, filteredSuggestion[newFocus]];
      }
      setTagsState(newTags);
      getSelectedTags && getSelectedTags(newTags);
      clearTags && clearTags(false);
      setInputValue('');
      setFocusedSuggestion(0);
    }
    if (key === KEYS.DOWN_ARROW) {
      const newFocus = focusedSuggestion + 1;
      setFocusedSuggestion(newFocus > filteredSuggestion?.length - 1 ? 0 : newFocus);
    }
    if (key === KEYS.UP_ARROW) {
      const newFocus = focusedSuggestion - 1;
      setFocusedSuggestion(focusedSuggestion < 1 ? filteredSuggestion?.length - 1 : newFocus);
    }
    if (key === KEYS.BACKSPACE && !val && isFocus) {
      removeTag(tagsState.length - 1);
    }
    if (key === KEYS.ESCAPE) {
      setInputValue('');
    }
  };

  const handleChange = (e: React.FormEvent<HTMLInputElement>) => {
    const value = e.currentTarget.value;
    handleInputChange && handleInputChange(value);
    setInputValue(value);
  };

  const onClickSuggestion = (suggestion: TagElement) => {
    if (checkDuplicateTag(suggestion)) {
      setErrorMessage(t('EXISTING_VALUE_ERROR', { value: suggestion.email }));
      return;
    }
    const newTags: TagElement[] = [...tagsState, suggestion];
    setTagsState(newTags);
    getSelectedTags && getSelectedTags(newTags);
    clearTags && clearTags(false);
    setInputValue('');
    inputRef.current?.focus();
  };

  const handleFocus = () => {
    setFocus(true);
  };

  const handleBlur = () => {
    setFocus(false);
  };

  const handleMouseOut = () => {
    setFocusedSuggestion(-1);
  };

  const checkDuplicateTag = (newTag: TagElement) => {
    return tagsState.find((tag) => tag.email === newTag.email);
  };

  const inputEventHandlers = {
    handleChange,
    handleKeyDown,
    handleBlur,
    handleFocus
  };

  return (
    <div className="input-tag-container">
      <div className="input-tag">
        <ul className="input-tag-list">
          <Tag tags={tagsState} removeTag={removeTag} />
          <InputTag
            {...inputEventHandlers}
            value={inputValue}
            isLoading={isLoading}
            customLoader={customLoader}
            inputRef={inputRef}
            readonly={readonly}
          />
        </ul>
      </div>
      <Col xs={12} className="px-0 text-error">
        {isFocus && errorMessage && <FormFeedback className="text-danger">{errorMessage}</FormFeedback>}
      </Col>
      {inputValue.length >= 2 && !errorMessage && (
        <Suggestions
          handleMouseOut={handleMouseOut}
          focusedSuggestion={focusedSuggestion}
          suggestions={filteredSuggestion}
          onClickSuggestion={onClickSuggestion}
          message={t('NO_RESULT')}
        />
      )}
    </div>
  );
};

export default TagsInput;
