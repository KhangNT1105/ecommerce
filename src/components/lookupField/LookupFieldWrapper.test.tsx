import React from 'react';
import { shallow } from 'enzyme';

import { testSnapshots } from 'utils/test';
import LookupFieldWrapper from 'components/lookupField/LookupFieldWrapper';
import { actions } from 'components/tagsInput/TagsInput.store';

jest.mock('react-router-dom', () => ({
  useHistory: () => ({
    push: jest.fn()
  })
}));

const defaultProps = {
  fieldName: '',
  single: false
};

describe('<LookupFieldWrapper />', () => {
  testSnapshots(LookupFieldWrapper, [
    {
      props: defaultProps,
      description: 'renders LookupFieldWrapper correctly'
    }
  ]);
});

describe('Test logic function', () => {
  let wrapper: any;
  const getState = jest.fn(() => ({}));
  const setState = jest.fn();
  const thunk = actions.setParams({ q: 'example1' });

  beforeEach(() => {
    jest.clearAllMocks();

    wrapper = shallow(<LookupFieldWrapper {...defaultProps} />);
  });

  test('should handle Input Change when input text', () => {
    thunk({ setState, getState } as any);
    const expectedResult = {
      data: 'example1',
      isSearching: true,
      params: {
        q: 'example1',
        sort: null
      }
    };

    const findLookupFieldTags = wrapper.find('LookupFieldTags');
    findLookupFieldTags.props().handleInputChange('example1');

    expect(setState).toHaveBeenCalledWith({
      params: {
        q: 'example1'
      }
    });
  });
});
