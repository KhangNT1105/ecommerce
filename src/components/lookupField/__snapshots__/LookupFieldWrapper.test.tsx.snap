// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LookupFieldWrapper /> snapshots renders LookupFieldWrapper correctly 1`] = `
<div
  className="input-tag-container"
>
  <div
    className="input-tag"
  >
    <ul
      className="input-tag-list"
    >
      <li
        className="input-box"
      >
        <input
          className=""
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          type="text"
          value=""
        />
      </li>
    </ul>
  </div>
  <div
    className="px-0 text-error col-12"
  />
  <div
    className="suggestions-container"
  />
  <div
    className="px-0 text-error col-12"
  />
</div>
`;
